cmd/apiserver/app/server/util.go

```diff
@@ -17,48 +17,154 @@ limitations under the License.
-		return nil, err
+		return nil, nil, err
 	}
-
 	// server configuration options
-	if err := s.SecureServingOptions.MaybeDefaultWithSelfSignedCerts(
-		s.GenericServerRunOptions.AdvertiseAddress.String(),
-	); err != nil {
-		return nil, err
+	if err := s.SecureServingOptions.MaybeDefaultWithSelfSignedCerts(s.GenericServerRunOptions.AdvertiseAddress.String()); err != nil {
+		return nil, nil, err
 	}
+	// NOTE: in k8s 1.7, this will take the explicit codec as input
 	genericConfig := genericapiserver.NewConfig()
 	if err := s.GenericServerRunOptions.ApplyTo(genericConfig); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	if err := s.SecureServingOptions.ApplyTo(genericConfig); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-
+	// this MUST be done after secure so we have a valid loopbackClientConfig
 	if err := s.InsecureServingOptions.ApplyTo(genericConfig); err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+	if inCluster {
+		if err := s.AuthenticationOptions.ApplyTo(genericConfig); err != nil {
+			return nil, nil, err
+		}
+	}
+	// TODO: add support for audit log options
+	// TODO: add support for OpenAPI config
+	genericConfig.SwaggerConfig = genericapiserver.DefaultSwaggerConfig()
+	// TODO: investigate if we need metrics unique to service catalog, but take defaults for now
+	genericConfig.EnableMetrics = true
+	// TODO: add support to default these values in build
+	serviceCatalogVersion := version.Get()
+	genericConfig.Version = &serviceCatalogVersion
+
+	// FUTURE: use protobuf for communication back to itself?
+	client, err := internalclientset.NewForConfig(genericConfig.LoopbackClientConfig)
+	if err != nil {
+		glog.Errorf("Failed to create clientset for service catalog self-communication: %v", err)
+		return nil, nil, err
+	}
+	sharedInformers := informers.NewSharedInformerFactory(client, 10*time.Minute)
+
+	scConfig := &serviceCatalogConfig{
+		client:          client,
+		sharedInformers: sharedInformers,
 	}
+	if inCluster {
+		inClusterConfig, err := restclient.InClusterConfig()
+		if err != nil {
+			glog.Errorf("Failed to get kube client config: %v", err)
+			return nil, nil, err
+		}
+		inClusterConfig.GroupVersion = &schema.GroupVersion{}
+
+		kubeClient, err := kubeclientset.NewForConfig(inClusterConfig)
+		if err != nil {
+			glog.Errorf("Failed to create clientset interface: %v", err)
+			return nil, nil, err
+		}
 
-	// glog.V(4).Info("Setting up authn (disabled)")
-	// need to figure out what's throwing the `missing clientCA file` err
-	/*
-		if _, err := genericConfig.ApplyDelegatingAuthenticationOptions(serverOptions.AuthenticationOptions); err != nil {
-			glog.Infoln(err)
-			return err
+		kubeSharedInformers := kubeinformers.NewSharedInformerFactory(kubeClient, 10*time.Minute)
+
+		// TODO: we need upstream to package AlwaysAdmit, or stop defaulting to it!
+		// NOTE: right now, we only run admission controllers when on kube cluster.
```

Comment:
because the only admission controller we have defined right now requires a kube cluster to communicate in order to actually admit requests.  the `KubernetesNamespaceLifecycle` admission controller will reject incoming requests to create incoming resources that are namespace-scoped if it cannot communicate to kube.